File: tests/result-channel-key.test.py
#!/usr/bin/env python3
"""Unit tests for src/result_channel_key.py — the per-channel pull path
for task-result files in `results/`.

Twin of tests/result-channel-key.test.ts. Same invariants, same shape.

Run: python3 tests/result-channel-key.test.py
Exit code: 0 on pass, 1 on fail.
"""

import json
import sys
import unittest
from pathlib import Path

REPO = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(REPO / "src"))

from result_channel_key import (  # noqa: E402
    sanitize_key,
    result_filename,
    parse_result_filename,
    result_belongs_to,
    phone_call_key,
)


class TestSanitizeKey(unittest.TestCase):
    def test_passes_safe_input(self):
        self.assertEqual(sanitize_key("1234567890123456789"), "1234567890123456789")
        self.assertEqual(sanitize_key("CA1234abcd"), "CA1234abcd")
        self.assertEqual(sanitize_key("local-voice"), "local-voice")
        self.assertEqual(sanitize_key("foo_bar-baz"), "foo_bar-baz")

    def test_collapses_unsafe_chars(self):
        self.assertEqual(sanitize_key("a/b"), "a-b")
        self.assertEqual(sanitize_key("a.b"), "a-b")
        self.assertEqual(sanitize_key("a b"), "a-b")
        self.assertEqual(sanitize_key("../etc/passwd"), "---etc-passwd")

    def test_empty_falls_back_to_unknown(self):
        self.assertEqual(sanitize_key(""), "unknown")
        self.assertEqual(sanitize_key(None), "unknown")
        self.assertEqual(sanitize_key("   "), "unknown")
        # All-unsafe → all dashes (not 'unknown' — input wasn't empty).
        self.assertEqual(sanitize_key("..."), "---")


class TestResultFilename(unittest.TestCase):
    def test_builds_scoped_form(self):
        self.assertEqual(
            result_filename("1234567890123456789", "task-plugin-voice-1700000000"),
            "1234567890123456789.task-plugin-voice-1700000000.txt",
        )
        self.assertEqual(
            result_filename("CA1234abcd", "task-phone-1700000000"),
            "CA1234abcd.task-phone-1700000000.txt",
        )


class TestParseResultFilename(unittest.TestCase):
    def test_splits_scoped_form(self):
        self.assertEqual(
            parse_result_filename("1234567890123456789.task-plugin-voice-1700000000.txt"),
            ("1234567890123456789", "task-plugin-voice-1700000000"),
        )
        self.assertEqual(
            parse_result_filename("CA1234abcd.task-phone-1700000000"),
            ("CA1234abcd", "task-phone-1700000000"),
        )

    def test_returns_none_for_legacy_flat(self):
        self.assertEqual(
            parse_result_filename("task-1700000000.txt"), (None, "task-1700000000")
        )
        self.assertEqual(
            parse_result_filename("task-plugin-voice-1700000000.txt"),
            (None, "task-plugin-voice-1700000000"),
        )

    def test_returns_none_for_non_task(self):
        self.assertEqual(
            parse_result_filename("voice-1700000000.txt"), (None, "voice-1700000000")
        )
        self.assertEqual(
            parse_result_filename("proactive-1700000000.txt"),
            (None, "proactive-1700000000"),
        )


class TestResultBelongsTo(unittest.TestCase):
    def test_claims_scoped_match(self):
        self.assertTrue(
            result_belongs_to(
                "1234567890123456789.task-foo.txt", "1234567890123456789"
            )
        )
        self.assertTrue(result_belongs_to("CA123.task-phone-1.txt", "CA123"))

    def test_rejects_different_key(self):
        self.assertFalse(
            result_belongs_to(
                "1234567890123456789.task-foo.txt", "9999999999"
            )
        )

    def test_rejects_legacy_flat(self):
        self.assertFalse(result_belongs_to("task-1700000000.txt", "1234567890123456789"))
        self.assertFalse(
            result_belongs_to("task-plugin-voice-1700000000.txt", "local-voice")
        )

    def test_rejects_non_task(self):
        self.assertFalse(result_belongs_to("voice-1700000000.txt", "local-voice"))
        self.assertFalse(result_belongs_to("proactive-1700000000.txt", "anything"))
        # Scoped form whose payload isn't a task-* file.
        self.assertFalse(
            result_belongs_to(
                "1234567890123456789.proactive-foo.txt", "1234567890123456789"
            )
        )

    def test_rejects_atomic_write_temp_suffixes(self):
        """Partial-write race: a writer's atomic-write temp file
        (``<key>.task-X.txt.tmp``, ``.sending``, ``.partial``, etc.) must
        NEVER match — picking it up would inject a half-written body and
        orphan the rename target. The scan loops also gate on
        ``endswith('.txt')``, but lock the invariant at the helper too."""
        key = "1234567890123456789"
        temp_suffixes = [
            "1234567890123456789.task-plugin-voice-1700000000.txt.tmp",
            "1234567890123456789.task-plugin-voice-1700000000.txt.partial",
            "1234567890123456789.task-plugin-voice-1700000000.txt.sending",
            "1234567890123456789.task-plugin-voice-1700000000.txt.swp",
            "1234567890123456789.task-plugin-voice-1700000000.txt.lock",
            "1234567890123456789.task-plugin-voice-1700000000.txt~",
            "1234567890123456789.task-plugin-voice-1700000000.sending",
            "1234567890123456789.task-plugin-voice-1700000000.tmp",
            "1234567890123456789.task-plugin-voice-1700000000.partial",
            # dotfile prefix (vim swap, atomic-write idioms)
            ".1234567890123456789.task-plugin-voice-1700000000.txt",
        ]
        for f in temp_suffixes:
            self.assertFalse(
                result_belongs_to(f, key),
                f"result_belongs_to should reject {f} (partial-write temp)",
            )

    def test_still_matches_canonical_txt(self):
        """Sanity-check: canonical `.txt` form still matches — the
        temp-suffix rejection didn't accidentally over-reject."""
        self.assertTrue(
            result_belongs_to(
                "1234567890123456789.task-plugin-voice-1700000000.txt",
                "1234567890123456789",
            )
        )


class TestExistingConsumersDoNotMatch(unittest.TestCase):
    """Load-bearing invariant. A scoped filename must NOT match any
    existing consumer's filter (specific task_id existsSync / `task-*`
    glob / startswith). Replay each consumer's actual pattern to lock
    that in."""

    SCOPED = "1234567890123456789.task-plugin-voice-1700000000.txt"
    SCOPED_BASE = "1234567890123456789.task-plugin-voice-1700000000"

    def test_pending_replies_lookup(self):
        # discord/telegram/slack bridges: result_file = RESULTS_DIR / f"{task_id}.txt"
        # where task_id is an id THEY tracked. A scoped filename's task_id
        # is the full prefixed string, which is never a tracked id.
        tracked_ids = ["task-1700000001", "task-plugin-voice-1700000000"]
        for tid in tracked_ids:
            self.assertNotEqual(
                f"{tid}.txt",
                self.SCOPED,
                f"pending id {tid} would match scoped filename",
            )

    def test_agent_api_glob(self):
        # agent-api.py: results_dir.glob("task-*.txt")
        # Equivalent: name starts with 'task-' and ends with '.txt'.
        self.assertFalse(self.SCOPED.startswith("task-"))

    def test_task_bridge_voice_guard(self):
        # task-bridge.ts: file.startsWith('voice-')
        self.assertFalse(self.SCOPED.startswith("voice-"))

    def test_task_bridge_task_guards(self):
        # task-bridge.ts: file.startsWith('task-') for dedup + offline forward
        self.assertFalse(self.SCOPED.startswith("task-"))

    def test_task_bridge_chat_guard(self):
        # task-bridge.ts: taskId.startsWith('task-chat-')
        self.assertFalse(self.SCOPED_BASE.startswith("task-chat-"))


class TestTypedKeyConstructors(unittest.TestCase):
    """Per-consumer prefixes — writer + consumer MUST go through the same
    typed function so the keys agree. Prevents cross-consumer namespace
    collisions when a future consumer ID format overlaps with an existing one."""

    def test_phone_call_key_prefixes_phone(self):
        self.assertEqual(phone_call_key("CA1234abcd"), "phone-CA1234abcd")

    def test_typed_keys_sanitize_input(self):
        self.assertEqual(phone_call_key("../etc"), "phone----etc")

    def test_typed_keys_fallback_on_empty(self):
        self.assertEqual(phone_call_key(None), "phone-unknown")

    def test_typed_keys_round_trip_through_result_filename(self):
        key = phone_call_key("CA1234abcd")
        fname = result_filename(key, "task-1700000000")
        self.assertEqual(fname, "phone-CA1234abcd.task-1700000000.txt")
        self.assertTrue(result_belongs_to(fname, key))


class TestCrossLanguageParityFixture(unittest.TestCase):
    """Asserts the SAME shared fixture the TS twin loads
    (tests/result-channel-key.test.ts) — one input/output table, asserted in
    both languages, so the Python impl (src/result_channel_key.py) and the TS
    impl (src/result-channel-key.ts) can never silently drift. Change an impl →
    the matching row fails in the OTHER language's test too. (Lucy on #1595.)
    """

    @classmethod
    def setUpClass(cls):
        path = REPO / "tests" / "fixtures" / "result-channel-key.parity.json"
        cls.fx = json.loads(path.read_text())

    def test_sanitize_key(self):
        for c in self.fx["sanitizeKey"]:
            self.assertEqual(sanitize_key(c["in"]), c["out"], f"sanitize_key({c['in']!r})")

    def test_result_filename(self):
        for c in self.fx["resultFilename"]:
            self.assertEqual(
                result_filename(c["channelKey"], c["taskId"]), c["out"], repr(c)
            )

    def test_parse_result_filename(self):
        for c in self.fx["parseResultFilename"]:
            key, task_id = parse_result_filename(c["in"])
            self.assertEqual(key, c["key"], f"key for {c['in']}")
            self.assertEqual(task_id, c["taskId"], f"taskId for {c['in']}")

    def test_result_belongs_to(self):
        for c in self.fx["resultBelongsTo"]:
            self.assertEqual(
                result_belongs_to(c["filename"], c["channelKey"]), c["out"], repr(c)
            )

    def test_phone_call_key(self):
        for c in self.fx["phoneCallKey"]:
            self.assertEqual(phone_call_key(c["in"]), c["out"], repr(c))


if __name__ == "__main__":
    unittest.main()

File: tests/result-channel-key.test.ts
/**
 * Unit tests for src/result-channel-key.ts — the per-channel pull path
 * for task-result files in `results/`.
 *
 * Load-bearing invariant: a `<channel-key>.task-{id}.txt` filename must
 * NOT match any existing consumer's filter (startsWith('task-') / glob
 * 'task-*.txt' / specific-task_id existsSync). We assert that here against
 * the actual patterns those consumers use, so a future change to one of
 * the helper functions can't silently re-open the blast radius.
 *
 * Run: tsx --test tests/result-channel-key.test.ts
 */

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import {
	sanitizeKey,
	resultFilename,
	parseResultFilename,
	resultBelongsTo,
	phoneCallKey,
} from '../src/result-channel-key.js';

describe('sanitizeKey', () => {
	it('passes through filename-safe input', () => {
		assert.equal(sanitizeKey('1234567890123456789'), '1234567890123456789');
		assert.equal(sanitizeKey('CA1234abcd'), 'CA1234abcd');
		assert.equal(sanitizeKey('local-voice'), 'local-voice');
		assert.equal(sanitizeKey('foo_bar-baz'), 'foo_bar-baz');
	});

	it('collapses unsafe chars to dashes', () => {
		assert.equal(sanitizeKey('a/b'), 'a-b');
		assert.equal(sanitizeKey('a.b'), 'a-b');
		assert.equal(sanitizeKey('a b'), 'a-b');
		assert.equal(sanitizeKey('../etc/passwd'), '---etc-passwd');
	});

	it('returns "unknown" for empty / falsy input', () => {
		assert.equal(sanitizeKey(''), 'unknown');
		assert.equal(sanitizeKey(null), 'unknown');
		assert.equal(sanitizeKey(undefined), 'unknown');
		assert.equal(sanitizeKey('   '), 'unknown');
		// All-unsafe → all dashes (NOT 'unknown' — the input wasn't empty).
		assert.equal(sanitizeKey('...'), '---');
	});
});

describe('resultFilename', () => {
	it('builds <key>.<task-id>.txt', () => {
		assert.equal(
			resultFilename('1234567890123456789', 'task-plugin-voice-1700000000'),
			'1234567890123456789.task-plugin-voice-1700000000.txt',
		);
		assert.equal(
			resultFilename('CA1234abcd', 'task-phone-1700000000'),
			'CA1234abcd.task-phone-1700000000.txt',
		);
	});
});

describe('parseResultFilename', () => {
	it('splits the scoped form', () => {
		assert.deepEqual(
			parseResultFilename('1234567890123456789.task-plugin-voice-1700000000.txt'),
			['1234567890123456789', 'task-plugin-voice-1700000000'],
		);
		assert.deepEqual(
			parseResultFilename('CA1234abcd.task-phone-1700000000'),
			['CA1234abcd', 'task-phone-1700000000'],
		);
	});

	it('returns [null, base] for the legacy flat form', () => {
		assert.deepEqual(parseResultFilename('task-1700000000.txt'), [null, 'task-1700000000']);
		assert.deepEqual(parseResultFilename('task-plugin-voice-1700000000.txt'), [
			null,
			'task-plugin-voice-1700000000',
		]);
	});

	it('returns [null, base] for non-task files (voice-, proactive-)', () => {
		assert.deepEqual(parseResultFilename('voice-1700000000.txt'), [null, 'voice-1700000000']);
		assert.deepEqual(parseResultFilename('proactive-1700000000.txt'), [
			null,
			'proactive-1700000000',
		]);
	});
});

describe('resultBelongsTo', () => {
	it('claims the scoped form for a matching key', () => {
		assert.equal(
			resultBelongsTo('1234567890123456789.task-foo.txt', '1234567890123456789'),
			true,
		);
		assert.equal(resultBelongsTo('CA123.task-phone-1.txt', 'CA123'), true);
	});

	it('rejects a different channel key', () => {
		assert.equal(
			resultBelongsTo('1234567890123456789.task-foo.txt', '9999999999'),
			false,
		);
	});

	it('rejects the legacy flat form (owned by delegating consumer, not by scan)', () => {
		assert.equal(resultBelongsTo('task-1700000000.txt', '1234567890123456789'), false);
		assert.equal(resultBelongsTo('task-plugin-voice-1700000000.txt', 'local-voice'), false);
	});

	it('rejects non-task files', () => {
		assert.equal(resultBelongsTo('voice-1700000000.txt', 'local-voice'), false);
		assert.equal(resultBelongsTo('proactive-1700000000.txt', 'anything'), false);
		assert.equal(resultBelongsTo('1234567890123456789.proactive-foo.txt', '1234567890123456789'), false);
	});

	// Partial-write race: a writer's atomic-write temp file (`<key>.task-X.txt.tmp`,
	// `.sending`, `.partial`, etc.) must NEVER match — picking it up would inject a
	// half-written body and orphan the rename target. The scan loops also gate on
	// `.endsWith('.txt')`, but lock the invariant at the helper too.
	it('rejects atomic-write temp suffixes (partial-write race)', () => {
		const KEY = '1234567890123456789';
		const tempSuffixes = [
			'1234567890123456789.task-plugin-voice-1700000000.txt.tmp',
			'1234567890123456789.task-plugin-voice-1700000000.txt.partial',
			'1234567890123456789.task-plugin-voice-1700000000.txt.sending',
			'1234567890123456789.task-plugin-voice-1700000000.txt.swp',
			'1234567890123456789.task-plugin-voice-1700000000.txt.lock',
			'1234567890123456789.task-plugin-voice-1700000000.txt~',
			'1234567890123456789.task-plugin-voice-1700000000.sending',
			'1234567890123456789.task-plugin-voice-1700000000.tmp',
			'1234567890123456789.task-plugin-voice-1700000000.partial',
			'.1234567890123456789.task-plugin-voice-1700000000.txt', // dotfile prefix (vim swap, atomic-write idioms)
		];
		for (const f of tempSuffixes) {
			assert.equal(
				resultBelongsTo(f, KEY),
				false,
				`resultBelongsTo should reject ${f} (partial-write temp)`,
			);
		}
	});

	// Sanity-check: the canonical `.txt` form for the same key still matches —
	// the temp-suffix rejection didn't accidentally over-reject.
	it('still matches the canonical .txt form', () => {
		assert.equal(
			resultBelongsTo(
				'1234567890123456789.task-plugin-voice-1700000000.txt',
				'1234567890123456789',
			),
			true,
		);
	});
});

// --- The load-bearing invariant: existing consumers don't see new files ---
// These tests replay the EXACT filter patterns the existing consumers use
// against a sample scoped filename. If any of these flips to `true`, the
// new namespace has leaked into a consumer's path and we've broken the
// blast-radius guarantee.
describe('existing consumers do NOT match the scoped namespace', () => {
	const SCOPED = '1234567890123456789.task-plugin-voice-1700000000.txt';
	const SCOPED_BASE = '1234567890123456789.task-plugin-voice-1700000000';

	it('discord-bridge / telegram-bridge / slack-bridge pending_replies lookup', () => {
		// All three bridges do: result_file = RESULTS_DIR / f"{task_id}.txt"
		// where task_id is the id they themselves tracked when writing the
		// task. A scoped filename's `task_id` would be the full prefixed
		// string, which is not a tracked id — so the existsSync miss.
		// Equivalent to: no pending_replies key matches SCOPED.
		const pending: Record<string, boolean> = {
			'task-1700000001': true,
			'task-plugin-voice-1700000000': true, // a hypothetical tracked id
		};
		// The bridge would look up `${task_id}.txt`; SCOPED doesn't equal any
		// `${tracked}.txt`.
		for (const tracked of Object.keys(pending)) {
			assert.notEqual(`${tracked}.txt`, SCOPED, `pending id ${tracked} would match scoped filename`);
		}
	});

	it('agent-api.py task-* glob does NOT match', () => {
		// Python: results_dir.glob("task-*.txt")
		// Equivalent JS: startsWith('task-') && endsWith('.txt')
		assert.equal(SCOPED.startsWith('task-'), false);
	});

	it('task-bridge.ts file.startsWith("voice-") guard does NOT match', () => {
		assert.equal(SCOPED.startsWith('voice-'), false);
	});

	it('task-bridge.ts file.startsWith("task-") guards (dedup-marker + voice-offline forward) do NOT match', () => {
		assert.equal(SCOPED.startsWith('task-'), false);
	});

	it('task-bridge.ts task-chat- guard does NOT match', () => {
		assert.equal(SCOPED_BASE.startsWith('task-chat-'), false);
	});

	// Note: task-bridge.ts has an UNCONDITIONAL fallthrough at line 682
	// (`if (result) { ... onResult(result) ... }`) that fires for any
	// non-empty .txt file when the voice client is connected. The narrow
	// design accepts this: voice-agent and the pull-side plugin / phone are
	// different surfaces, and in practice the active pull-side / phone
	// process will read-and-delete the scoped file before voice-agent's
	// 2s poll claims it. See PR body's verification section for the full
	// trade-off discussion.
});

// Per-consumer prefixes — writer + consumer MUST go through the same typed
// function so the keys agree. Prevents cross-consumer namespace collisions
// when a future consumer ID format overlaps with an existing one.
describe('typed key constructors', () => {
	it('phoneCallKey prefixes with `phone-`', () => {
		assert.equal(phoneCallKey('CA1234abcd'), 'phone-CA1234abcd');
	});

	it('typed keys sanitize input', () => {
		assert.equal(phoneCallKey('../etc'), 'phone----etc');
	});

	it('typed keys fall back on empty / falsy input', () => {
		assert.equal(phoneCallKey(null), 'phone-unknown');
	});

	it('typed keys round-trip through resultFilename + resultBelongsTo', () => {
		const key = phoneCallKey('CA1234abcd');
		const fname = resultFilename(key, 'task-1700000000');
		assert.equal(fname, 'phone-CA1234abcd.task-1700000000.txt');
		assert.equal(resultBelongsTo(fname, key), true);
	});
});

// ── Cross-language parity ──────────────────────────────────────────────────
// The SAME shared fixture the Python twin (tests/result-channel-key.test.py)
// loads. One input/output table, asserted in both languages, so the TS impl
// (src/result-channel-key.ts) and the Python impl (src/result_channel_key.py)
// can never silently drift. Change an impl → the matching row fails in the
// OTHER language's test too. (Lucy's flag on PR #1595.)
describe('cross-language parity fixture', () => {
	const fixture = JSON.parse(
		readFileSync(new URL('./fixtures/result-channel-key.parity.json', import.meta.url), 'utf8'),
	) as Record<string, any[]>;

	it('sanitizeKey matches the shared fixture', () => {
		for (const c of fixture.sanitizeKey) {
			assert.equal(sanitizeKey(c.in), c.out, `sanitizeKey(${JSON.stringify(c.in)})`);
		}
	});

	it('resultFilename matches the shared fixture', () => {
		for (const c of fixture.resultFilename) {
			assert.equal(resultFilename(c.channelKey, c.taskId), c.out, JSON.stringify(c));
		}
	});

	it('parseResultFilename matches the shared fixture', () => {
		for (const c of fixture.parseResultFilename) {
			const [key, taskId] = parseResultFilename(c.in);
			assert.equal(key, c.key, `key for ${c.in}`);
			assert.equal(taskId, c.taskId, `taskId for ${c.in}`);
		}
	});

	it('resultBelongsTo matches the shared fixture', () => {
		for (const c of fixture.resultBelongsTo) {
			assert.equal(resultBelongsTo(c.filename, c.channelKey), c.out, JSON.stringify(c));
		}
	});

	it('phoneCallKey matches the shared fixture', () => {
		for (const c of fixture.phoneCallKey) {
			assert.equal(phoneCallKey(c.in), c.out, JSON.stringify(c));
		}
	});
});
